internal/controllers/core/cmd/controller.go

```diff
@@ -103,15 +103,26 @@ func (c *Controller) reconcile(ctx context.Context, name types.NamespacedName) e
 		return nil
 	}
 
-	lastEventTime, err := c.restartManager.lastEventTime(ctx, cmd.Spec.RestartOn)
+	lastRestartEventTime, err := c.triggerManager.lastEventTime(ctx, cmd.Spec.RestartOn)
+	if err != nil {
+		return err
+	}
+	lastStartEventTime, err := c.triggerManager.lastEventTime(ctx, cmd.Spec.StartOn)
 	if err != nil {
 		return err
 	}
 
 	proc, ok := c.procs[name]
+	// there is already a proc (which may or may not be running) for this cmd
 	if ok {
-		needsRestart := !equality.Semantic.DeepEqual(proc.spec, cmd.Spec) ||
-			lastEventTime.After(proc.lastEventTime)
+		// any change to the spec means we should restart the process to pick up the changes
+		specChanged := !equality.Semantic.DeepEqual(proc.spec, cmd.Spec)
+		// a new restart event happened
+		restartOnTriggered := lastRestartEventTime.After(proc.lastRestartEventTime)
+		// a new start event happened *and* the existing proc is finished
```

Comment:
Hmmmm...I'm concerned this API isn't going to behave like you expect.

A good rule of thumb with this kind of distributed API is that it needs to be idempotent - if the controller accidentally gets the event twice or if the timing is a bit off, that's OK.

So we really don't want an API that behaves differently gets ignored if the cmd is running and does an action if the cmd is completed. I think that's a recipe for lots of unresolvable problems and race conditions.